### PR TITLE
Fix use-after-free in Stats.

### PR DESCRIPTION
**Pull request by jplevyak**

### CODE REVIEW

File: source/extensions/common/wasm/wasm.cc [hunk 1]
```diff
@@ -2841,6 +2840,12 @@ std::shared_ptr<Wasm> getOrCreateThreadLocalWasm(Wasm& base_wasm, absl::string_v
   auto wasm = getThreadLocalWasmPtr(base_wasm.vm_id());
   if (wasm) {
     auto root_context = wasm->start();
+    // NB: we are reusing this VM with the new configuration so we need to replace the plugin which
+    // contains a Scope& which is in a ListenerImpl which may be deleted. The Scope has no prefix
+    // (""), and does nothing but provide a layer of caching and forward to the underlying
+    // Stats::ThreadLocalStoreImpl.
+    // TODO(jplevyak): this will go away after https://github.com/envoyproxy/envoy-wasm/issues/258
```
Comment (mandarjog) [inline source/extensions/common/wasm/wasm.cc]: What does base_wasm.plugin() return? Does this only affect wasm or also null sandbox?

Can you add a bit of explanation? I see this one extra line and a change of name?

**Comment by mandarjog** [comment]
@kyessenov can you test this fix with your stress test so as to save big roundtrips ?

I am deploying proxy with this fix in my large cluster.
gcr.io/mixologist-142215/proxyv2:1.4.0-asanfix

File: source/extensions/common/wasm/wasm.cc (hunk 1 above)
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: base_wasm.plugin() returns a plugin with a valid Scope.  As commented, this is going to change when we move the VM declarations out of the the plugin configuration.  At that point we will have control of the lifetime and we will not have a plugin which is generic to the Wasm and instead we will only have the plugin in the Context (as it should be).

**Comment by mandarjog** [comment]
I used the above build that includes this fix and got 2 segfaults.
``` c++
[Envoy (Epoch 1)] [2019-10-20 00:28:04.247][287][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Segmentation fault, suspect faulting address 0x5095f8
[Envoy (Epoch 1)] [2019-10-20 00:28:04.247][287][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[Envoy (Epoch 1)] [2019-10-20 00:28:04.247][287][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 1195ecbc70d058a50f424d5f68f0f07e269ef331/1.12.0-dev/Clean/RELEASE/BoringSSL
[Envoy (Epoch 1)] [2019-10-20 00:28:04.247][287][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: __restore_rt [0x7f413310d890]
[Envoy (Epoch 1)] [2019-10-20 00:28:04.254][287][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #1: Envoy::Extensions::Common::Wasm::recordMetricHandler() [0x104a086]
[Envoy (Epoch 1)] [2019-10-20 00:28:04.261][287][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: Envoy::Extensions::Common::Wasm::Null::Plugin::Stats::PluginRootContext::report() [0xbcb6b6]
[Envoy (Epoch 1)] [2019-10-20 00:28:04.268][287][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #3: std::__1::__function::__func<>::operator()() [0xc46ed9]

2019-10-20T00:28:26.428371Z	error	Epoch 1 exited with error: signal: segmentation fault (core dumped)
2019-10-20T00:28:26.428402Z	info	No more active epochs, terminating
```

**Comment by mandarjog** [comment]
```c++

[Envoy (Epoch 0)] [2019-10-20 00:00:52.655][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Segmentation fault, suspect faulting address 0x4d9edb0
[Envoy (Epoch 0)] [2019-10-20 00:00:52.655][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[Envoy (Epoch 0)] [2019-10-20 00:00:52.655][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 1195ecbc70d058a50f424d5f68f0f07e269ef331/1.12.0-dev/Clean/RELEASE/BoringSSL
[Envoy (Epoch 0)] [2019-10-20 00:00:52.655][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: __restore_rt [0x7fb3762d1890]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.663][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #1: Envoy::Config::GrpcMuxImpl::sendDiscoveryRequest() [0x1cdcb09]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.669][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: Envoy::Config::GrpcMuxImpl::GrpcMuxWatchImpl::~GrpcMuxWatchImpl() [0x1cdee87]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.676][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #3: Envoy::Config::GrpcMuxImpl::GrpcMuxWatchImpl::~GrpcMuxWatchImpl() [0x1cdef1e]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.682][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #4: Envoy::Config::GrpcSubscriptionImpl::~GrpcSubscriptionImpl() [0x1ccab48]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.688][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #5: Envoy::Secret::SdsApi::~SdsApi() [0x1b736ba]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.695][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #6: Envoy::Extensions::TransportSockets::Tls::ContextConfigImpl::~ContextConfigImpl() [0x1adc00d]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.701][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #7: Envoy::Extensions::TransportSockets::Tls::ClientContextConfigImpl::~ClientContextConfigImpl() [0x1add053]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.707][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #8: Envoy::Extensions::TransportSockets::Tls::ClientSslSocketFactory::~ClientSslSocketFactory() [0x1ad9222]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.713][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #9: Envoy::Upstream::TransportSocketMatcherImpl::~TransportSocketMatcherImpl() [0x1ce4f4d]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.719][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #10: Envoy::Upstream::ClusterInfoImpl::~ClusterInfoImpl() [0x1cb2973]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.725][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #11: Envoy::Upstream::ClusterInfoImpl::~ClusterInfoImpl() [0x1cb29ce]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.731][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #12: Envoy::Upstream::HostDescriptionImpl::~HostDescriptionImpl() [0xd297ad]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.738][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #13: Envoy::Upstream::HostImpl::~HostImpl() [0xd30c07]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.744][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #14: std::__1::__shared_ptr_pointer<>::__on_zero_shared() [0xd2a199]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.750][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #15: Envoy::Upstream::HostSetImpl::~HostSetImpl() [0xfe2506]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.756][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #16: Envoy::Upstream::HostSetImpl::~HostSetImpl() [0x1cb257e]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.762][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #17: Envoy::Upstream::PrioritySetImpl::~PrioritySetImpl() [0xd280db]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.769][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #18: Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::~ClusterEntry() [0x1b941de]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.775][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #19: Envoy::Upstream::ClusterManagerImpl::ThreadLocalClusterManagerImpl::~ThreadLocalClusterManagerImpl() [0x1b91afd]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.781][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #20: Envoy::ThreadLocal::InstanceImpl::shutdownThread() [0x1b0add9]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.787][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #21: Envoy::Server::InstanceImpl::terminate() [0x1b14211]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.793][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #22: Envoy::Server::InstanceImpl::run() [0x1b1a872]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.800][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #23: Envoy::MainCommonBase::run() [0xceb230]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.806][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #24: main [0xcea252]
[Envoy (Epoch 0)] [2019-10-20 00:00:52.806][19][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #25: __libc_start_main [0x7fb375eefb97]
2019-10-20T00:00:53.118587Z	error	Epoch 0 exited with error: signal: segmentation fault (core dumped)
```

**Pull request merged**